### PR TITLE
Rewrite in TypeScript

### LINKED ISSUE #8: Rewrite with TypeScript

### PR DESCRIPTION
**Pull request by Cryrivers**
Fixes #8, supersedes #16 

Basically I followed the guidance in #8 and achieved almost 100% type coverage.

cc @crosscompile @jaredpalmer

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,21 +34,28 @@
...
+    "@babel/cli": "^7.1.2",
+    "@babel/core": "^7.1.2",
+    "@babel/plugin-proposal-class-properties": "^7.1.0",
+    "@babel/plugin-proposal-object-rest-spread": "^7.0.0",
+    "@babel/plugin-transform-runtime": "^7.1.0",
+    "@babel/preset-env": "^7.1.0",
+    "@babel/preset-react": "^7.0.0",
+    "@babel/preset-typescript": "^7.1.0",
+    "@babel/runtime": "^7.1.2",
+    "@types/lodash": "^4.14.117",
```
Comment (jaredpalmer) [inline package.json]: Is this needed?

**Comment by jaredpalmer** [comment]
This is rad! 

Quick question though: Is it possible to even consume this externally? Wouldn't the person need to pull in the exact files of your `react-cache` and `react` types overrides too? Idk. Could be talking out of my ass, but have a hunch.

File: package.json (hunk 1 above)
Comment (Cryrivers) [reply package.json]: Yes, iirc `lodash` was used somewhere in the library. Therefore the type declaration is needed. It's purely used in compile-type.

**Comment by Cryrivers** [comment]
> Is it possible to even consume this externally?
Of course, yes! the type declaration files for TypeScript users are generated in `dist`.

> Wouldn't the person need to pull in the exact files of your react-cache and react types overrides too? 
By right they wouldn't. I need to verify this, because I thought I might need to add those `@types/*` packages as `dependencies` instead of `devDependencies`.

**Comment by jaredpalmer** [comment]
Seems a tad shaky right now. Let’s wait until the official types arrive

**Comment by Cryrivers** [comment]
Sure thing. For `@types/react` with hooks supported, let's wait for DefinitelyTyped/DefinitelyTyped#30036.

For the type for `react-cache`, I guess we need to include it anyway as there's no official typing files. I just ported the flowtype from `react` repo.

**Comment by Cryrivers** [comment]
@jaredpalmer since official React 16.7 declaration has been released (DefinitelyTyped/DefinitelyTyped#30057). Could you help review this PR?

File: package.json (hunk 1 above)
Comment (Cryrivers) [reply package.json]: Removed.

**Comment by Cryrivers** [comment]
CI is unable to run on master due to this PR. Let me try to fix it.

**Pull request merged**